test/scripts/run_cuda_memcheck.py

```diff
@@ -0,0 +1,101 @@
+#!/usr/bin/env python
+
+"""This script runs cuda-memcheck on the specified unit test. Each test case
+is run in its isolated process with a timeout so that:
+1) different test cases won't influence each other, and
+2) in case of hang, the script would still finish in a finite amount of time.
+The output will be written to a log file result.log
+
+Example usage:
+    python run_cuda_memcheck.py ../test_torch.py 600
+
+Note that running cuda-memcheck could be very slow.
+"""
+
+import asyncio
+import torch
+import multiprocessing
+import argparse
+import subprocess
+import tqdm
+import re
+
+ALL_TESTS = []
+GPUS = torch.cuda.device_count()
+
+# parse arguments
+parser = argparse.ArgumentParser(description="Run isolated cuda-memcheck on unit tests")
+parser.add_argument('filename', help="the python file for a test, such as test_torch.py")
+parser.add_argument('timeout', type=int, help='kill the test if it does not terminate in a certain amount of seconds')
+parser.add_argument('--ignore', nargs='+', default=['cudaErrorInvalidDeviceFunction'],
+                    help='list of regex of the failures not interested, default to ["cudaErrorInvalidDeviceFunction"], '
+                         'because cublas does not run error-free under cuda-memcheck')
+parser.add_argument('--nproc', type=int, default=multiprocessing.cpu_count(),
```

Comment:
Follow-up thought: should the default be the number of gpus? Combined with the below comment, that ensures each gpu is only running one test at at time, preventing interfernce.

Reply:
I don't this is a good idea, for the following reason:

1. The test is CPU bound, each process will occupy one CPU core at 100%, and almost 0% GPU utilization. The reason to distribute across GPUs is mainly due to the availability of CUDA memory. And it is very slow and usually, on a server, there are more CPU cores than the number of GPUs.

2. Different tests are running on different processes. If I understand correctly, different processes have different page tables, therefore different virtual addresses, making the interference of each other very unlikely.

Reply:
OK. Now that users can set it we can have our cake and eat it, too, here.